### PR TITLE
CART-898 CART: add UCX support to DAOS.

### PR DESCRIPTION
**Pull request by jgmoore-or**
CaRT code is changed to enable UCX protocol strings.

Build script (site_scons/components/\_\_init\_\_.py) is changed to build Mercury using the system UCX installation.

Tested with MOFED: MLNX_OFED_LINUX-5.5-1.0.3.2 (earlier versions of MOFED have earlier UCX releases and may not work).

UCX requires the user's memory limits be set in /etc/security/limits.conf:

username soft memlock unlimited
username hard memlock unlimited

Client must have the following env vars set:

export OFI_INTERFACE=ib0
export OFI_DOMAIN=mlx5_0:1,mlx5_1:1

export CRT_PHY_ADDR_STR=same as in server config 

For example: ucx+rc_x,ud_x

will enable both reliable-connected and unreliable datagram transports.

The server config should also specify both devices as OFI_DOMAIN for each engine.

The use of OFI_ environment variable names is a temporary work-around.  Once the control plane changes to support UCX are in place, it is assumed that non-OFI-based names will be enabled.

Control plane currently does not allow UCX protocol strings, so this PR includes interim changes to two control plane files to work around this limitation.  This PR should not be landed until issue is fixed in the control plane code and the control plane changes are removed from this PR.

Until a patch update PR is landed in daos-stack/mercury, utils/build.config must specify patching mercury with:

https://raw.githubusercontent.com/daos-stack/mercury/95e9ac9bd706bc256a6f984c4a847fa264a3975a/mercury_ucx_parse_addr_change.patch

### CODE REVIEW

**Review by johannlombardi — CHANGES_REQUESTED** [review]
A few changes are required. I suspect that more is needed on the control plane side to address the unit test failure and validate "ucx+" as a valid provider.

File: src/cart/crt_hg.h [hunk 1]
```diff
@@ -40,9 +40,19 @@ enum crt_na_type {
 	CRT_NA_OFI_PSM2		= 4,
 	CRT_NA_OFI_TCP_RXM	= 5,
 	CRT_NA_OFI_CXI		= 6,
```
Comment (johannlombardi) [inline src/cart/crt_hg.h:42]: Maybe add a CRT_NA_OFI_LAST = CRT_NA_OFI_CX, here

File: src/cart/crt_hg.h [hunk 2]
```diff
@@ -52,11 +62,19 @@ crt_prov_str_to_na_type(const char *prov_str);
...
+	return (na_type >= CRT_NA_UCX_RC) &&
+	       (na_type < CRT_NA_COUNT);
+}
+
 static inline bool
 crt_na_type_is_ofi(int na_type)
 {
 	return (na_type >= CRT_NA_OFI_SOCKETS) &&
-	       (na_type < CRT_NA_OFI_COUNT);
+	       (na_type <= CRT_NA_OFI_CXI);
```
Comment (johannlombardi) [inline src/cart/crt_hg.h:77]: and reuse CRT_NA_OFI_LAST there.

**Review by johannlombardi — APPROVED** [review]

**Comment by johannlombardi** [comment]
Could someone else please inspect it?

**Comment by mjmac** [comment]
@kjacque: Would you please take a look and work with @jgmoore-or to get this aligned with your UCX work?

File: site_scons/components/__init__.py [hunk 3]
```diff
@@ -172,7 +172,19 @@ def define_mercury(reqs):
...
+    reqs.define("UCX", libs=['ucp'])
+
+    if reqs.check_component('UCX'):
+        UCX = '-DNA_USE_UCX=ON ' \
+        '-DUCX_INCLUDE_DIR=/usr/include '\
+        '-DUCP_LIBRARY=/usr/lib64/libucp.so '\
+        '-DUCS_LIBRARY=/usr/lib64/libucs.so '\
+        '-DUCT_LIBRARY=/usr/lib64/libuct.so '
+    else:
+        UCX = ""
```
Comment (frostedcmos) [inline site_scons/components/__init__.py:186]: should we do explicit NA_USE_UCX=OFF  in this case?

File: src/cart/crt_context.c [hunk 4]
```diff
@@ -221,13 +221,15 @@ crt_context_provider_create(crt_context_t *crt_ctx, int provider)
...
-	rc = crt_hg_get_addr(ctx->cc_hg_ctx.chc_hgcla,
-			     ctx->cc_self_uri, &uri_len);
-	if (rc != 0) {
-		D_ERROR("ctx_hg_get_addr() failed; rc: %d.\n", rc);
-		D_RWLOCK_UNLOCK(&crt_gdata.cg_rwlock);
-		crt_context_destroy(ctx, true);
-		D_GOTO(out, rc);
+	if (crt_is_service() || !crt_na_type_is_ucx(provider)) {
+		rc = crt_hg_get_addr(ctx->cc_hg_ctx.chc_hgcla,
+				     ctx->cc_self_uri, &uri_len);
```
Comment (frostedcmos) [inline src/cart/crt_context.c:226]: how is self_uri set for ucx?
crt_self_uri_get() and crt_rank_uri_get() APIs rely on this value to be set; this API is then used in number of tests

File: src/cart/crt_hg.c [hunk 5]
```diff
@@ -664,11 +714,13 @@ crt_hg_class_init(int provider, int idx, hg_class_t **ret_hg_class)
 		D_GOTO(out, rc = -DER_HG);
 	}
 
-	rc = crt_hg_get_addr(hg_class, addr_str, &str_size);
-	if (rc != 0) {
-		D_ERROR("crt_hg_get_addr() failed, rc: %d.\n", rc);
-		HG_Finalize(hg_class);
-		D_GOTO(out, rc = -DER_HG);
+	if (crt_is_service() || !crt_na_type_is_ucx(provider)) {
+		rc = crt_hg_get_addr(hg_class, addr_str, &str_size);
```
Comment (frostedcmos) [inline src/cart/crt_hg.c:718]: same question here as above, what about the UCX case? addr_str is supposed to be stored in the context, and later used in various places

File: site_scons/components/__init__.py (hunk 3 above)
Comment (jolivier23) [reply site_scons/components/__init__.py:186]: that's probably safer depending on the mercury default behavior

File: site_scons/components/__init__.py [hunk 6]
```diff
@@ -172,7 +172,19 @@ def define_mercury(reqs):
         MERCURY_DEBUG = '-DMERCURY_ENABLE_DEBUG=OFF '
     retriever = \
         GitRepoRetriever('https://github.com/mercury-hpc/mercury.git',
-                         True)
+	True)
+
+    reqs.define("UCX", libs=['ucp'])
```
Comment (jolivier23) [inline site_scons/components/__init__.py:177]: Every other component uses lower case.  My personal preference would be to be consistent and just call it ucx but this is a minor point.

File: site_scons/components/__init__.py [hunk 7]
```diff
@@ -172,7 +172,19 @@ def define_mercury(reqs):
...
+
+    if reqs.check_component('UCX'):
+        UCX = '-DNA_USE_UCX=ON ' \
+        '-DUCX_INCLUDE_DIR=/usr/include '\
+        '-DUCP_LIBRARY=/usr/lib64/libucp.so '\
+        '-DUCS_LIBRARY=/usr/lib64/libucs.so '\
+        '-DUCT_LIBRARY=/usr/lib64/libuct.so '
```
Comment (jolivier23) [inline site_scons/components/__init__.py:184]: You should also append 'UCX' (or 'ucx' if you follow suggestion above) to libs list so it gets added to requires= line in the mercury definition.   It should be listed as a requirement of mercury if it's available.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
some minor issues.

File: site_scons/components/__init__.py (hunk 7 above)
Comment (jolivier23) [reply site_scons/components/__init__.py:184]: Add libs.append('ucx') so the dependence is explicit on line 210

**Review by jolivier23 — CHANGES_REQUESTED** [review]

**Review by johannlombardi — APPROVED** [review]

**Comment by johannlombardi** [comment]
Please inspect. I would like to land ASAP.

**Review by jolivier23 — APPROVED** [review]

**Review by kjacque — APPROVED** [review]
The control plane changes were the only blocker for me. With them removed I'm OK with landing.

**Comment by kjacque** [comment]
I'm still working on control plane support for UCX, so as things are UCX cannot be enabled without disabling control plane provider validation.

**Pull request merged**